server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -532,20 +533,20 @@ protected void doRun() {
                     prohibitCustomRoutingOnDataStream(docWriteRequest, metadata);
                     prohibitAppendWritesInBackingIndices(docWriteRequest, metadata);
                     docWriteRequest.routing(metadata.resolveWriteIndexRouting(docWriteRequest.routing(), docWriteRequest.index()));
-                    docWriteRequest.process();
 
                     final Index concreteIndex = docWriteRequest.getConcreteWriteIndex(ia, metadata);
                     if (addFailureIfIndexIsClosed(docWriteRequest, concreteIndex, i, metadata)) {
                         continue;
                     }
                     IndexRouting indexRouting = concreteIndices.routing(concreteIndex);
+                    docWriteRequest.process(indexRouting);
                     int shardId = docWriteRequest.route(indexRouting);
```

Comment:
It'd be possible to merge `process` and `route`. Or to have `process` call the `indexRouting` and return the shard id. I didn't do that because I kind of like having `route` not modify the request.

Reply:
++